src/smart-account.ts

```diff
@@ -0,0 +1,314 @@
+import {
+    ethers,
+    BlockTag,
+    Signer,
+    TypedDataDomain,
+    TypedDataField,
+    getAddress,
+    SigningKey,
+    BytesLike,
+} from "ethers";
+import { Provider } from "./provider";
+import { EIP712Signer } from "./signer";
+import { Wallet } from "./wallet";
+import { EIP712_TX_TYPE, DEFAULT_GAS_PER_PUBDATA_LIMIT, serializeEip712 } from "./utils";
+import { Address, TransactionResponse, TransactionRequest, TransactionLike, Eip712Meta } from "./types";
+
+/**
+ * The custom signing method is used to sign transactions with a custom logic
+ * @param transaction - transaction to sign
+ * @param defaultSigner - default signer to sign transaction
+ * @param keys - private keys to sign transaction
+ * @returns signed transaction
+ */
+type CustomSigningMethod = (
+    transaction: TransactionRequest,
+    defaultSigner: ethers.Signer,
+    keys: string[] | string | ethers.SigningKey,
+) => Promise<string>;
+
+export class SmartAccount extends EIP712Signer implements Signer {
+    readonly address: string;
+    provider: Provider;
+    readonly customSigningMethod?: CustomSigningMethod;
+    readonly privateKeys: string[] | string | ethers.SigningKey;
+    private defaultSigner: ethers.Signer;
```

Comment:
Consider making use of explicit `Wallet` instance instead. Storing the `Wallet` instance as a class member (e.g., this.wallet) would make its presence more explicit and allow for **its direct use** within `SmartAccount` methods. For example, 

```
private wallet: Wallet;

const primaryKey = Array.isArray(privateKeys) ? privateKeys[0] : privateKeys;
this.wallet = new Wallet(primaryKey, provider); 
super(this.wallet, network.then((n) => Number(n.chainId)));

this.defaultSigner = this.wallet;
```

The `signTransaction`, `signMessage`, and `signTypedData` methods use `this.defaultSigner` for signing operations. To maintain consistency and leverage the `Wallet` functionalities, these methods could utilize the `Wallet` instance created during initialization. This would ensure that all signing operations are routed through the `Wallet`.

Example:
```
async signTransaction(tx: TransactionRequest): Promise<string> {
        if (this.customSigningMethod) {
            return this.customSigningMethod(tx, this.wallet, this.privateKeys);
        } else {
            return this.wallet.signTransaction(tx);
        }
    }
```

Reply:
Yeah, we could do that but now that I think of it, maybe we should only initialise the `Wallet` instance when `signingMethod` is not passed in the constructor, which means we're using the default signing. 

Although in the other cases, we'd need not just a `customSigningMethod` but a custom signer with methods to `signTransaction`, `signMessage`, and `signTypedData`